src/reader.rs

```diff
@@ -82,17 +111,127 @@ pub fn find(needle: &ReadUntil, buffer: &str, eof: bool) -> Option<(usize, usize
                 None
             }
         }
-        &ReadUntil::Any(ref any) => {
-            for read_until in any {
-                if let Some(pos_tuple) = find(&read_until, buffer, eof) {
-                    return Some(pos_tuple);
-                }
-            }
+    }
+}
+
+pub struct Match<T> {
+    begin: usize,
+    end: usize,
+    pub interest: T,
+}
+
+impl<T> Match<T> {
+    fn new(begin: usize, end: usize, obj: T) -> Self {
+        Self {
+            begin,
+            end,
+            interest: obj,
+        }
+    }
+}
+
+pub trait Needle {
+    type Interest;
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>>;
+}
+
+impl Needle for str {
+    type Interest = String;
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>> {
+        buffer.find(self).and_then(|pos| {
+            Some(Match::new(
+                pos,
+                pos + self.len(),
+                buffer[..pos].to_string(),
+            ))
+        })
+    }
+}
+
+impl Needle for String {
+    type Interest = String;
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>> {
+        buffer.find(self).and_then(|pos| {
+            Some(Match::new(
+                pos,
+                pos + self.len(),
+                buffer[..pos].to_string(),
+            ))
+        })
+    }
+}
+
+impl Needle for Regex {
+    type Interest = (String, String);
+
+    fn find(&self, buffer: &str, eof: bool) -> Option<Match<Self::Interest>> {
```

Comment:
why is there a `Needle::find` function for regex here and also a function `find` which also handles regex?

Reply:
If you mean this one I just didn't delete it.
 https://github.com/philippkeller/rexpect/blob/0212cd6ff069e5633ea8e6d4dad8212f88ef6464/src/reader.rs#L86

but if you mean `Regex::find(&self, buffer)`, here we just call original find method `regex::Regex::find`.

Reply:
yes I meant the `pub fn find` function. Nice that it can be deleted. All clear then!